### PR TITLE
[Backport] 8305896: Alternative full GC forwarding

### PR DESCRIPTION
**Pull request by mmyxym**
Summary: Introduce UseAltGCForwarding to support alternative full GC forwarding.
	 Now support G1 and parallelGC. The code accords to 8305896 and https://github.com/openjdk/lilliput-jdk17u/

Test Plan: CICD

Reviewed-by: yude.lin, yifeng.jin

Issue: https://github.com/dragonwell-project/dragonwell11/issues/683

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1FullGCOopClosures.inline.hpp [hunk 1]
```diff
@@ -73,24 +75,34 @@ template <class T> inline void G1AdjustClosure::adjust_pointer(T* p) {
...
+             obj->mark_raw()->must_be_preserved(obj) || // Will be restored by PreservedMarksSet
+             (UseBiasedLocking && obj->has_bias_pattern_raw()), // Will be restored by BiasedLocking
+             "Must have correct prototype or be preserved, obj: " PTR_FORMAT ", mark: " PTR_FORMAT ", prototype: " PTR_FORMAT,
+             p2i(obj), p2i(obj->mark_raw()), p2i(markOopDesc::prototype_for_object(obj)));
+    }
     return;
   }
+  if (ALT_FWD) {
+    forwardee = SlidingForwarding::forwardee<true>(obj);
+  }
```
Comment (linade) [inline src/hotspot/share/gc/g1/g1FullGCOopClosures.inline.hpp:95]: This seems unnecessarily complicated. Can't we put it like:
```
if (ALT_FWD) {
  xxx
} else {
  xxx
}
```

**Review by leveretconey — COMMENTED** [review]
Since only G1 and PS are currently supported, would it be best for us to check during JVM initialization and disable the option if G1 or PS are not selected？

**Comment by mmyxym** [comment]
> Since only G1 and PS are currently supported, would it be best for us to check during JVM initialization and disable the option if G1 or PS are not selected？

The option will not affect other GCs

**Comment by mmyxym** [comment]
> Since only G1 and PS are currently supported, would it be best for us to check during JVM initialization and disable the option if G1 or PS are not selected？

The option will not affect other GCs.

**Review by leveretconey — APPROVED** [review]
LGTM

File: src/hotspot/share/gc/g1/g1FullGCOopClosures.inline.hpp (hunk 1 above)
Comment (mmyxym) [reply src/hotspot/share/gc/g1/g1FullGCOopClosures.inline.hpp]: updated

**Review by linade — APPROVED** [review]
lgtm

**Pull request merged**